File: 0121 - Best Time to Buy and Sell Stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        vector<int> arr = profit(prices);
        return arr[0];
    }

private:
    vector<int> profit(vector<int>& prices){
        int profit = 0;
        vector<int> result;
        result.push_back(profit);
        if(prices.size() < 2)
            return result;
        
        int buy_index = 0, sell_index = 1;
        int buy = prices[0], sell;
        
        for(int i = 1; i < prices.size(); i++){
            sell = prices[i];
            if(profit < sell - buy){
                profit = sell - buy;
                sell_index = i;
            }
            if(sell < buy){
                buy = sell;
                buy_index = i;
            }
        }

        result[0] = profit;
        result.push_back(buy_index);
        result.push_back(sell_index);
        return result;
    }
};

File: 0121 - Best Time to Buy and Sell Stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        min_list, max_list = [], []
        
        mi, ma = prices[0], prices[-1]
        result = 0
        for index, price in enumerate(prices):
            mi = min(mi, price)
            min_list.append(mi)

            ma = max(ma, prices[len(prices) - 1 - index])
            max_list.append(ma)

        for index, mi in enumerate(min_list):
            result = max(result, max_list[len(max_list) - 1 - index] - mi)

        return result